File: stack/BOJ/2493/2493.py
def main():   
    n = int(input())
    str = input()
    towers = [int(i) for i in str.split(' ')]
    res = []
    stack = []
    for i in range(len(towers)):
        height = towers[i]
        while (stack and stack[-1][1] < height):
            stack.pop()
        if stack:
            res.append(stack[-1][0])
        else:
            res.append(0)
        stack.append([i + 1, height])
    
    for i in res:
        print(i , end=' ')


if __name__ == "__main__":
    main()

File: stack/BOJ/2493/2493.cpp
#include <bits/stdc++.h>
using namespace std;
#define X first
#define Y second

int N;
stack<pair<int,int>> tower; // 높이 값과 인덱스 저장

int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);

  cin >> N;
  for (int i = 1; i <= N; i++) {
    int height; // 입력된 값 저장
    cin >> height;
    
    // 입력된 값보다 작은 숫자들은 정답 가능성이 없으므로 제거한다.
    while (!tower.empty() && tower.top().X < height) tower.pop();
    if (tower.empty())  cout << "0 "; // 스택이 비어있다면 0을 출력
    else cout << tower.top().Y << " "; // 비어있지 않다면 top에 있는 값을 출력
    tower.push({height, i}); // 입력된 값과 인덱스는 스택에 저장한다.     
  }
  return 0;
}
